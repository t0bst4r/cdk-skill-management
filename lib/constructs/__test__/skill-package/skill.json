{
  "manifest": {
    "manifestVersion": "1.0",
    "publishingInformation": {
      "distributionCountries": [],
      "locales": {
        "de-DE": {
          "description": "My Personal Smart Home",
          "examplePhrases": [],
          "keywords": [
            "Smart Home",
            "Lights",
            "Smart Devices"
          ],
          "name": "Test-Smart-Home-Skill",
          "summary": "My Personal Smart Home"
        }
      },
      "testingInstructions": "Sample Testing Instructions."
    },
    "permissions": [
      {
        "name": "alexa::async_event:write"
      }
    ],
    "apis": {
      "smartHome": {
        "protocolVersion": "3",
        "endpoint": {
          "uri": "//override"
        }
      }
    }
  }
}