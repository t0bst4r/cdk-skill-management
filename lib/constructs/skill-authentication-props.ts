import {CfnSkill} from 'aws-cdk-lib/alexa-ask';
import {IStringParameter} from 'aws-cdk-lib/aws-ssm';
import {ISecret} from 'aws-cdk-lib/aws-secretsmanager';

/**
 * Type alias for Alexa Skill authentication configuration.
 */
export type SkillAuthenticationConfiguration = CfnSkill.AuthenticationConfigurationProperty;

/**
 * Properties for configuring the authentication properties for Skill Management.
 */
export interface SkillAuthenticationProps {
  /** Authentication configuration for the Alexa Skill. */
  readonly authenticationConfiguration?: SkillAuthenticationConfiguration;
  /** StringParameter holding the authentication configuration for the Alexa Skill. */
  readonly authenticationConfigurationParameter?: IStringParameter;
  /** Secret holding the authentication configuration for the Alexa Skill. */
  readonly authenticationConfigurationSecret?: ISecret;
}
