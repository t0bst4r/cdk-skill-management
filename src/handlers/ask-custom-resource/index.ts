import type {AskSdkCall, SkillAuthenticationConfiguration} from '../../../lib';
import type {CdkCustomResourceHandler} from 'aws-lambda';
import {StandardSmapiClientBuilder} from 'ask-smapi-sdk';
import {GetParameterCommand, SSMClient} from '@aws-sdk/client-ssm';
import {GetSecretValueCommand, SecretsManagerClient} from '@aws-sdk/client-secrets-manager';

export const handler: CdkCustomResourceHandler = async event => {
  const sdkCall: AskSdkCall | undefined = getSdkCall(event.RequestType, event.ResourceProperties);
  if (!sdkCall) {
    console.log(`No SDK call configured for ${event.RequestType}`);
    return {};
  }

  const config = await getAuthenticationConfig(event.ResourceProperties);

  const client = new StandardSmapiClientBuilder().withRefreshTokenConfig(config).client();
  const genericClient = client as object as Record<
    string,
    (...args: Array<unknown>) => Promise<Record<string, unknown>>
  >;
  const action = sdkCall.action in genericClient ? genericClient[sdkCall.action] : undefined;
  if (!action) {
    throw new Error(`ASK SDK does not provide an ${sdkCall.action} action`);
  }
  const boundAction = action.bind(client);
  const result = await boundAction(...sdkCall.parameters).catch(e => {
    if (e.response) {
      if (e.response.message) {
        throw new Error(e.response.message + '\n' + JSON.stringify(e.response, undefined, 2));
      } else {
        throw new Error(JSON.stringify(e.response, undefined, 2));
      }
    }
    throw e;
  });
  return {Data: result};
};

function getSdkCall(type: 'Create' | 'Update' | 'Delete', properties: Record<string, unknown>): AskSdkCall | undefined {
  let result: string | undefined;
  if (type === 'Create') {
    result = (properties.create ?? properties.update) as string | undefined;
  } else if (type === 'Update') {
    result = properties.update as string | undefined;
  } else {
    result = properties.delete as string | undefined;
  }
  return result ? JSON.parse(result) : undefined;
}

async function getAuthenticationConfig(parameters: Record<string, unknown>): Promise<SkillAuthenticationConfiguration> {
  if (parameters.authenticationConfigurationSecret) {
    const secretsManager = new SecretsManagerClient({});
    const secret = await secretsManager.send(
      new GetSecretValueCommand({
        SecretId: parameters.authenticationConfigurationSecret as string,
      })
    );
    if (secret.SecretString) {
      return JSON.parse(secret.SecretString);
    } else {
      throw new Error(`Secret ${secret.SecretString} resolved to null`);
    }
  } else if (parameters.authenticationConfigurationParameter) {
    const ssm = new SSMClient({});
    const parameter = await ssm.send(
      new GetParameterCommand({
        Name: parameters.authenticationConfigurationParameter as string,
      })
    );
    if (parameter.Parameter?.Value) {
      return JSON.parse(parameter.Parameter.Value);
    } else {
      throw new Error(`Parameter ${parameters.authenticationConfigurationParameter} resolved to null`);
    }
  } else if (parameters.authenticationConfiguration) {
    return JSON.parse(parameters.authenticationConfiguration as string);
  } else {
    throw new Error('No authentication configuration provided');
  }
}
