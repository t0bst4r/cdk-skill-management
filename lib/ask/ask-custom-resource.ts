import {
  CustomResource,
  CustomResourceProvider,
  CustomResourceProviderRuntime,
  Duration,
  IResolvable,
  Lazy,
  RemovalPolicy,
  Resource,
  Stack,
} from 'aws-cdk-lib';
import {SkillAuthenticationConfiguration} from '../constructs/skill';
import {Construct} from 'constructs';
import {AskSdkCall} from './ask-sdk-call';
import * as path from 'path';

/**
 * Properties for configuring an Ask Custom Resource.
 */
export interface AskCustomResourceProps {
  /** Timeout for the custom resource. */
  readonly timeout?: Duration;
  /** Removal policy for the custom resource. */
  readonly removalPolicy?: RemovalPolicy;
  /** Authentication configuration for the Alexa Skill. */
  readonly authenticationConfiguration: SkillAuthenticationConfiguration | IResolvable;
  /** Action to perform on resource creation. */
  readonly onCreate?: AskSdkCall;
  /** Action to perform on resource update. */
  readonly onUpdate?: AskSdkCall;
  /** Action to perform on resource deletion. */
  readonly onDelete?: AskSdkCall;
}

/**
 * A custom CloudFormation resource for Alexa Skill Kit SDK calls.
 */
export class AskCustomResource extends Resource {
  private readonly provider: CustomResourceProvider;
  private readonly customResource: CustomResource;

  /**
   * Creates an instance of the Ask Custom Resource.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Ask Custom Resource properties.
   */
  constructor(scope: Construct, id: string, props: AskCustomResourceProps) {
    super(scope, id);

    const codeDir =
      path.extname(__filename) === '.ts'
        ? path.join(__dirname, '..', '..', 'dist', 'src', 'handlers', 'ask-custom-resource')
        : path.join(__dirname, '..', '..', 'src', 'handlers', 'ask-custom-resource');

    this.provider = CustomResourceProvider.getOrCreateProvider(this, 'Custom::ASK', {
      codeDirectory: codeDir,
      runtime: CustomResourceProviderRuntime.NODEJS_18_X,
    });

    this.customResource = new CustomResource(this, 'Resource', {
      resourceType: 'Custom::ASK',
      serviceToken: this.provider.serviceToken,
      pascalCaseProperties: true,
      removalPolicy: props.removalPolicy,
      properties: {
        authentication: props.authenticationConfiguration,
        create: props.onCreate ? this.encodeJson(props.onCreate) : undefined,
        update: props.onUpdate ? this.encodeJson(props.onUpdate) : undefined,
        delete: props.onDelete ? this.encodeJson(props.onDelete) : undefined,
      },
    });
  }

  /**
   * Gets the response field from the custom resource.
   * @param dataPath - The data path to retrieve from the response.
   * @returns The value of the response field.
   */
  public getResponseField(dataPath: string): string {
    return this.customResource.getAttString(dataPath);
  }

  private encodeJson(obj: unknown): string {
    return Lazy.uncachedString({produce: () => Stack.of(this).toJsonString(obj)});
  }
}
