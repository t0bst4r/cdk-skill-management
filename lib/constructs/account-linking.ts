import {Construct, IConstruct} from 'constructs';
import type {ISkill} from './skill';
import {AskCustomResource} from '../ask/ask-custom-resource';
import {SkillAuthenticationProps} from './skill-authentication-props';

/**
 * Represents a platform-specific authorization URL for account linking.
 */
export interface AccountLinkingPlatformAuthorizationUrl {
  /** The platform type */
  readonly platformType: 'iOS' | 'Android';
  /** The platform-specific authorization URL. */
  readonly platformAuthorizationUrl: string;
}

/**
 * Represents the request parameters for account linking.
 */
export interface AccountLinkingRequest {
  /** The type of account linking */
  readonly authenticationFlowType?: 'AUTH_CODE' | 'IMPLICIT';
  /** The authorization URL for account linking. */
  readonly authorizationUrl?: string;
  /** An array of domains for account linking. */
  readonly domains?: Array<string>;
  /** The client ID for account linking. **/
  readonly clientId?: string;
  /** An array of scopes for account linking. */
  readonly scopes?: Array<string>;
  /** The access token URL for account linking. */
  readonly accessTokenUrl?: string;
  /** The client secret for account linking. */
  readonly clientSecret?: string;
  /** The access token scheme. */
  readonly accessTokenScheme?: 'HTTP_BASIC' | 'REQUEST_BODY_CREDENTIALS';
  /** The default token expiration in seconds for account linking. */
  readonly defaultTokenExpirationInSeconds?: number;
  /** The reciprocal access token URL for account linking. */
  readonly reciprocalAccessTokenUrl?: string;
  /** An array of redirect URLs for account linking. */
  readonly redirectUrl?: Array<string>;
  /** An array of platform-specific authorization URLs for account linking. */
  readonly authorizationUrlsByPlatform?: Array<AccountLinkingPlatformAuthorizationUrl>;
  /** Indicates whether to skip account linking on enablement. */
  readonly skipOnEnablement?: boolean;
  /** Voice-forward account linking setting, either 'ENABLED' or 'DISABLED'. */
  readonly voiceForwardAccountLinking?: 'ENABLED' | 'DISABLED';
}

/**
 * Interface representing an Account Linking resource.
 */
export interface IAccountLinking extends IConstruct {}

/**
 * Properties for creating an Account Linking resource.
 */
export interface AccountLinkingProps extends SkillAuthenticationProps {
  /** The Alexa Skill for which account linking is configured. */
  readonly skill: ISkill;
  /** The request parameters for account linking. */
  readonly request: AccountLinkingRequest;
}

/**
 * Represents an Account Linking resource for an Alexa Skill.
 */
export class AccountLinking extends Construct implements IAccountLinking {
  private readonly resource: AskCustomResource;

  /**
   * Creates an instance of the Account Linking resource.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Account Linking properties.
   */
  constructor(scope: Construct, id: string, props: AccountLinkingProps) {
    super(scope, id);

    this.resource = new AskCustomResource(this, 'Default', {
      authenticationConfiguration: props.authenticationConfiguration,
      authenticationConfigurationSecret: props.authenticationConfigurationSecret,
      authenticationConfigurationParameter: props.authenticationConfigurationParameter,
      onUpdate: {
        action: 'updateAccountLinkingInfoV1',
        parameters: [
          props.skill.skillId,
          props.skill.skillStage,
          {
            accountLinkingRequest: {
              ...props.request,
              type: props.request.authenticationFlowType,
              authenticationFlowType: undefined,
            },
          },
        ],
      },
      onDelete: {
        action: 'deleteAccountLinkingInfoV1',
        parameters: [props.skill.skillId, props.skill.skillStage],
      },
    });
    this.resource.node.addDependency(props.skill);
  }
}
