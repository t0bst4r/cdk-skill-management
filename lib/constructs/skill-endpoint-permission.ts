import {Construct, IDependable} from 'constructs';
import {CfnPermission, IFunction} from 'aws-cdk-lib/aws-lambda';
import {ISkill, SkillType} from './skill';
import {AwsCustomResource, AwsCustomResourcePolicy, AwsSdkCall, PhysicalResourceId} from 'aws-cdk-lib/custom-resources';
import {PolicyStatement} from 'aws-cdk-lib/aws-iam';

/**
 * Properties for configuring a Skill Endpoint Permission.
 */
export interface SkillEndpointProps {
  /** The AWS Lambda function handler to configure the permission for. */
  readonly handler: IFunction;
}

/**
 * Interface representing a Skill Endpoint Permission.
 */
export interface ISkillEndpointPermission {
  /**
   * Configures a Skill Endpoint Permission for a specific skill.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skill - The skill to configure the permission for.
   * @returns An IDependable object representing the configured permission.
   */
  configureSkillId(scope: Construct, id: string, skill: ISkill): IDependable;
}

/**
 * Class for configuring and managing Skill Endpoint Permissions.
 */
export class SkillEndpointPermission extends Construct implements ISkillEndpointPermission {
  private readonly principals: Record<SkillType, string> = {
    [SkillType.CUSTOM]: 'alexa-appkit.amazon.com',
    [SkillType.SMART_HOME]: 'alexa-connectedhome.amazon.com',
  };

  private readonly handler: IFunction;
  private readonly permission: CfnPermission;
  private readonly policy: AwsCustomResourcePolicy;

  /**
   * Creates an instance of the Skill Endpoint Permission.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Skill Endpoint properties.
   */
  constructor(scope: Construct, id: string, props: SkillEndpointProps) {
    super(scope, id);

    this.permission = new CfnPermission(this, 'InitialSkillPermission', {
      functionName: props.handler.functionArn,
      principal: this.principals[SkillType.CUSTOM],
      action: 'lambda:InvokeFunction',
    });

    this.policy = AwsCustomResourcePolicy.fromStatements([
      new PolicyStatement({
        actions: ['lambda:RemovePermission', 'lambda:AddPermission'],
        resources: [props.handler.functionArn],
      }),
    ]);

    this.handler = props.handler;
  }

  /**
   * Configures a Skill Endpoint Permission for a specific skill.
   * This replaces the initially created permission with a new permission checking for the SkillId.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skill - The skill to configure the permission for.
   * @returns An IDependable object representing the configured permission.
   */
  public configureSkillId(scope: Construct, id: string, skill: ISkill): IDependable {
    const parent = new Construct(scope, id);
    new AwsCustomResource(parent, 'RemovePermission', {
      policy: this.policy,
      onCreate: this.removePermissionCall(parent, skill),
      onDelete: this.addPermissionCall(parent),
      installLatestAwsSdk: false,
    });
    new AwsCustomResource(parent, 'AddPermission', {
      policy: this.policy,
      onCreate: this.addPermissionCall(parent, skill),
      onDelete: this.removePermissionCall(parent, skill),
      installLatestAwsSdk: false,
    });
    parent.node.addDependency(this.permission, this.handler);
    return parent;
  }

  private removePermissionCall(scope: Construct, skill: ISkill): AwsSdkCall {
    return {
      service: 'Lambda',
      action: 'removePermission',
      parameters: {
        FunctionName: this.handler.functionArn,
        StatementId: this.permission.attrId,
      },
      physicalResourceId: PhysicalResourceId.of(`RemovePermission-${scope.node.addr}-${skill.skillId}`),
    };
  }

  private addPermissionCall(scope: Construct, skill?: ISkill): AwsSdkCall {
    return {
      service: 'Lambda',
      action: 'addPermission',
      parameters: {
        FunctionName: this.handler.functionArn,
        StatementId: this.permission.attrId,
        Principal: this.principals[skill?.skillType ?? SkillType.CUSTOM],
        Action: 'lambda:InvokeFunction',
        EventSourceToken: skill?.skillId,
      },
      physicalResourceId: PhysicalResourceId.of(`AddPermission-${scope.node.addr}-${skill?.skillId ?? 'no-skill'}`),
    };
  }
}
