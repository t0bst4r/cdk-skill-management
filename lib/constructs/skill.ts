import {IResource, Resource} from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CfnSkill} from 'aws-cdk-lib/alexa-ask';
import {Role, ServicePrincipal} from 'aws-cdk-lib/aws-iam';
import {Asset} from 'aws-cdk-lib/aws-s3-assets';
import {SkillAuthenticationProps} from './skill-authentication-props';
import {ISecret} from 'aws-cdk-lib/aws-secretsmanager';
import {IStringParameter} from 'aws-cdk-lib/aws-ssm';

/**
 * Enumeration for different Alexa Skill types.
 */
export enum SkillType {
  CUSTOM = 'CUSTOM',
  SMART_HOME = 'SMART_HOME',
}

/**
 * Interface representing an Alexa Skill.
 */
export interface ISkill extends IResource {
  /** The unique ID of the Alexa Skill. */
  readonly skillId: string;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
}

/**
 * Base class for the Alexa Skill construct.
 */
abstract class SkillBase extends Resource implements ISkill {
  /**
   * The unique ID of the Alexa Skill.
   */
  public abstract readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public abstract readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public abstract readonly skillType: SkillType;
}

/**
 * Type alias for Alexa Skill package overrides.
 */
export type SkillPackageOverrides = CfnSkill.OverridesProperty;

/**
 * Interface representing an Alexa Skill package.
 */
export interface SkillPackage {
  /** The asset associated with the Alexa Skill package. */
  readonly asset: Asset;
  /** Overrides for the Alexa Skill package. */
  readonly overrides?: SkillPackageOverrides;
}

/**
 * Properties for creating an Alexa Skill.
 */
export interface SkillProps extends SkillAuthenticationProps {
  /** The type of the Alexa Skill. */
  readonly skillType: SkillType;
  /** The stage of the Alexa Skill. */
  readonly skillStage: string;
  /** The vendor ID of the Alexa Skill. */
  readonly vendorId: string;
  /** The package information for the Alexa Skill. */
  readonly skillPackage: SkillPackage;
}

/**
 * Properties for creating an Alexa Skill Instance from Attributes.
 */
export interface SkillAttributes {
  /**
   * The unique ID of the Alexa Skill.
   */
  readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  readonly skillType: SkillType;
}

/**
 * Alexa Skill construct for managing an Alexa Skill via CloudFormation.
 */
export class Skill extends SkillBase {
  /**
   * Factory method to create an Alexa Skill from its ID.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param skillId - The ID of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromSkillId(scope: Construct, id: string, skillId: string): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = skillId;

      get skillType(): SkillType {
        throw new Error('Access to skillType is not supported when using fromSkillId');
      }

      get skillStage(): string {
        throw new Error('Access to skillStage is not supported when using fromSkillId');
      }
    })(scope, id);
  }

  /**
   * Factory method to create an Alexa Skill from its attributes.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param attributes - The attributes of the Alexa Skill.
   * @returns The Alexa Skill construct.
   */
  public static fromAttributes(scope: Construct, id: string, attributes: SkillAttributes): ISkill {
    return new (class extends SkillBase {
      readonly skillId: string = attributes.skillId;
      readonly skillType: SkillType = attributes.skillType;
      readonly skillStage: string = attributes.skillStage;
    })(scope, id);
  }

  private readonly packageRole: Role;
  private readonly resource: CfnSkill;

  /**
   * The unique ID of the Alexa Skill.
   */
  public readonly skillId: string;
  /**
   * The stage of the Alexa Skill.
   */
  public readonly skillStage: string;
  /**
   * The type of the Alexa Skill.
   */
  public readonly skillType: SkillType;

  /**
   * Creates an instance of the Alexa Skill construct.
   * @param scope - The construct scope.
   * @param id - The construct ID.
   * @param props - The Alexa Skill properties.
   */
  constructor(scope: Construct, id: string, props: SkillProps) {
    super(scope, id);

    const authPropsCount = [
      props.authenticationConfiguration,
      props.authenticationConfigurationParameter,
      props.authenticationConfigurationSecret,
    ].filter(it => !!it).length;

    if (authPropsCount !== 1) {
      throw new Error('Exactly one authentication configuration needs to be provided!');
    }

    this.packageRole = new Role(this, 'PackageRole', {
      assumedBy: new ServicePrincipal('alexa-appkit.amazon.com'),
    });

    const authentication: CfnSkill.AuthenticationConfigurationProperty =
      this.authenticationSecret(props.authenticationConfigurationSecret) ??
      this.authenticationParameter(props.authenticationConfigurationParameter) ??
      props.authenticationConfiguration!;

    this.resource = new CfnSkill(this, 'Default', {
      skillPackage: {
        s3Bucket: props.skillPackage.asset.s3BucketName,
        s3Key: props.skillPackage.asset.s3ObjectKey,
        s3BucketRole: this.packageRole.roleArn,
        overrides: {
          ...props.skillPackage.overrides,
          manifest: {
            ...props.skillPackage.overrides?.manifest,
            publishingInformation: {
              category: props.skillType,
              ...props.skillPackage.overrides?.manifest?.publishingInformation,
            },
          },
        },
      },
      vendorId: props.vendorId,
      authenticationConfiguration: authentication,
    });
    this.resource.node.addDependency(this.packageRole);
    this.resource.node.addDependency(props.skillPackage.asset);
    props.skillPackage.asset.grantRead(this.packageRole);

    this.skillId = this.resource.ref;
    this.skillStage = props.skillStage;
    this.skillType = props.skillType;
  }

  private authenticationSecret(secret: ISecret | undefined): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!secret) {
      return;
    }
    return {
      clientId: secret.secretValueFromJson('clientId').unsafeUnwrap(),
      clientSecret: secret.secretValueFromJson('clientSecret').unsafeUnwrap(),
      refreshToken: secret.secretValueFromJson('refreshToken').unsafeUnwrap(),
    };
  }

  private authenticationParameter(
    parameter: IStringParameter | undefined
  ): CfnSkill.AuthenticationConfigurationProperty | undefined {
    if (!parameter) {
      return;
    }
    return JSON.parse(parameter.stringValue);
  }
}
