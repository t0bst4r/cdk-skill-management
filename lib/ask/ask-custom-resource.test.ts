import {App, Stack} from 'aws-cdk-lib';
import {Template} from 'aws-cdk-lib/assertions';
import {Secret} from 'aws-cdk-lib/aws-secretsmanager';
import {AskCustomResource} from './ask-custom-resource';
import {Skill, SkillType} from '../constructs/skill';

test('AskCustomResource', () => {
  const app = new App();
  const stack = new Stack(app, 'TestStack');

  const skillCredentials = Secret.fromSecretNameV2(stack, 'SkillCredentials', '/Skill/Credentials');

  const skill = Skill.fromAttributes(stack, 'Skill', {
    skillId: 'My-Skill-Id',
    skillStage: 'development',
    skillType: SkillType.SMART_HOME,
  });
  new AskCustomResource(stack, 'AccountLinking', {
    authenticationConfigurationSecret: skillCredentials,
    onUpdate: {
      action: 'updateAccountLinkingInfoV1',
      parameters: [skill.skillId, skill.skillStage, {accountLinkingRequest: {my: 1, awesome: 2, props: 3}}],
    },
    onDelete: {
      action: 'deleteAccountLinkingInfoV1',
      parameters: [skill.skillId, skill.skillStage],
    },
  });

  expect(Template.fromStack(stack)).toMatchSnapshot();
});
