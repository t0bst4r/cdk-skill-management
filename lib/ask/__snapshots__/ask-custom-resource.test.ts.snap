// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AskCustomResource 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AccountLinkingC99A8611": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Authentication": {
          "Fn::Join": [
            "",
            [
              "{{resolve:secretsmanager:arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":secretsmanager:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":secret:/Skill/Credentials:SecretString:::}}",
            ],
          ],
        },
        "Delete": "{"action":"deleteAccountLinkingInfoV1","parameters":["My-Skill-Id","development"]}",
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomASKCustomResourceProviderHandlerF5F0D5CE",
            "Arn",
          ],
        },
        "Update": "{"action":"updateAccountLinkingInfoV1","parameters":["My-Skill-Id","development",{"accountLinkingRequest":{"my":1,"awesome":2,"props":3}}]}",
      },
      "Type": "Custom::ASK",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomASKCustomResourceProviderHandlerF5F0D5CE": {
      "DependsOn": [
        "CustomASKCustomResourceProviderRoleB79D76EE",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "b0abcd4386b0c765d5354f12fb62cd61d7ca5a1bce853bac1869d7c79aa33744.zip",
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomASKCustomResourceProviderRoleB79D76EE",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomASKCustomResourceProviderRoleB79D76EE": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
